File: scripts/source/convert_date/convert_date.exs
#!/usr/bin/env elixir

defmodule DateConverter do
  @month_map %{
    "January" => "01",
    "February" => "02",
    "March" => "03",
    "April" => "04",
    "May" => "05",
    "June" => "06",
    "July" => "07",
    "August" => "08",
    "September" => "09",
    "October" => "10",
    "November" => "11",
    "December" => "12"
  }

  def convert_date(date_string) do
    case String.split(date_string, " ") do
      [month, day_and_comma, year] ->
        day = String.trim_trailing(day_and_comma, ",")
        month_number = Map.get(@month_map, month)

        if month_number do
          {:ok, "#{year}-#{month_number}-#{String.pad_leading(day, 2, "0")}"}
        else
          {:error, "Invalid month name: #{month}"}
        end

      _ ->
        {:error, "Invalid date format. Expected format: 'Month Day, Year' (e.g., 'October 20, 1999')."}
    end
  end
end

defmodule CLI do
  def main(args) do
    case args do
      [date_string] ->
        case DateConverter.convert_date(date_string) do
          {:ok, converted_date} ->
            IO.puts(converted_date)

          {:error, message} ->
            IO.puts(:stderr, "Error: #{message}")
        end

      _ ->
        IO.puts(:stderr, "Usage: elixir convert_date.exs \"Month Day, Year\"")
    end
  end
end

# Entrypoint
CLI.main(System.argv())

File: scripts/source/convert-date/convert-date.lua
local DateConverter = {}

-- Month map for converting month names to numeric format
DateConverter.month_map = {
    January = "01",
    February = "02",
    March = "03",
    April = "04",
    May = "05",
    June = "06",
    July = "07",
    August = "08",
    September = "09",
    October = "10",
    November = "11",
    December = "12"
}

-- Function to convert the date
function DateConverter.convert_date(date_string)
    -- Split the input into components
    local month, day, year = date_string:match("(%a+)%s(%d+),%s(%d+)")
    if month and day and year then
        local month_number = DateConverter.month_map[month]
        if month_number then
            -- Format the date as YYYY-MM-DD
            return string.format("%s-%s-%02d", year, month_number, tonumber(day))
        else
            return nil, "Invalid month name: " .. month
        end
    else
        return nil, "Invalid date format. Expected format: 'Month Day, Year' (e.g., 'October 20, 1999')."
    end
end

-- CLI handling
local function main()
    -- Get the input date from the command-line arguments
    local args = {...}
    if #args == 1 then
        local input_date = args[1]
        local converted_date, err = DateConverter.convert_date(input_date)
        if converted_date then
            print("Converted date: " .. converted_date)
        else
            io.stderr:write("Error: " .. err .. "\n")
        end
    else
        io.stderr:write("Usage: lua convert_date.lua \"Month Day, Year\"\n")
    end
end

-- Run the script
main

